File: jar/src/main/java/jline/api/mam/Qbd_depproc_etaqa.java
/**
 * @file QBD departure process ETAQA truncation for FCFS discipline
 *
 * @since LINE 3.0
 */
package jline.api.mam;

import java.util.Map;

import jline.lib.smc.QBD_CR;
import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class Qbd_depproc_etaqa {
    private Qbd_depproc_etaqa() {}

    /**
     * Compute MAP departure process for MAP/MAP/1-FCFS via ETAQA truncation.
     */
    public static MatrixCell qbd_depproc_etaqa(MatrixCell MAPa, MatrixCell MAPs, int n) {
        int na = MAPa.get(0).getNumRows();
        int ns = MAPs.get(0).getNumRows();
        int lvlsz = ns * na;

        Matrix IA = Matrix.eye(na);
        Matrix IS = Matrix.eye(ns);
        Matrix F = MAPa.get(1).kron(IS);
        Matrix L = MAPa.get(0).kron(IS).add(IA.kron(MAPs.get(0)));
        Matrix B = IA.kron(MAPs.get(1));
        Matrix L0 = MAPa.get(0).kron(IS);

        Map<String, Matrix> qbdResult = QBD_CR.QBD_CR(B, L, F, null, null, null, null);
        Matrix R = qbdResult.get("R");
        if (R == null) {
            throw new RuntimeException("QBD_CR failed to compute R matrix");
        }

        Matrix G = L.add(R.mult(B)).neg().inv().mult(B);
        Matrix Lhat = F.add(L);
        Matrix Bbar = B.add(F.mult(G));
        Matrix Bhat = F.mult(G);

        int totalDim = (n + 1) * lvlsz;

        Matrix D0 = Matrix.zeros(totalDim, totalDim);
        D0.insertSubMatrix(0, 0, lvlsz, lvlsz, L0);
        if (n >= 1) {
            D0.insertSubMatrix(0, lvlsz, lvlsz, 2 * lvlsz, F);
        }

        for (int j = 1; j <= n; j++) {
            int rs = j * lvlsz;
            D0.insertSubMatrix(rs, rs, rs + lvlsz, rs + lvlsz, L);
            if (j < n) {
                D0.insertSubMatrix(rs, rs + lvlsz, rs + lvlsz, rs + 2 * lvlsz, F);
            }
        }

        if (n >= 1) {
            int rs = (n - 1) * lvlsz;
            D0.insertSubMatrix(rs, rs, rs + lvlsz, rs + lvlsz, Lhat);
        }

        Matrix D1 = Matrix.zeros(totalDim, totalDim);
        if (n >= 1) {
            int rsN = n * lvlsz;
            int csN1 = (n - 1) * lvlsz;
            D1.insertSubMatrix(rsN, csN1, rsN + lvlsz, csN1 + lvlsz, Bbar);
            D1.insertSubMatrix(rsN, rsN, rsN + lvlsz, rsN + lvlsz, Bhat);
        }

        for (int j = 1; j < n; j++) {
            int rs = j * lvlsz;
            int cs = (j - 1) * lvlsz;
            D1.insertSubMatrix(rs, cs, rs + lvlsz, cs + lvlsz, B);
        }

        MatrixCell result = new MatrixCell(2);
        result.set(0, D0);
        result.set(1, D1);
        return result;
    }
}

File: matlab/src/api/mam/qbd_depproc_etaqa.m
%{ @file qbd_depproc_etaqa.m
 %  @brief Constructs MAP departure process for MAP/MAP/1-FCFS via ETAQA truncation
 %
 %  @author LINE Development Team
%}

%{
 % @brief Builds a MAP approximation of the departure process for a MAP/MAP/1 FCFS queue
 %
 % @details
 % This function constructs a finite-state MAP {D0, D1} representation of the
 % departure process from a MAP/MAP/1 queue with FCFS discipline, using ETAQA
 % truncation at QBD level n. The resulting MAP captures both the SCV (squared
 % coefficient of variation) and autocorrelation structure of inter-departure
 % times.
 %
 % The QBD process is formed with forward matrix F = kron(MAPa{2}, I_ns),
 % local matrix L = krons(MAPa{1}, MAPs{1}), and backward matrix
 % B = kron(I_na, MAPs{2}). Levels 0..n-1 are represented explicitly, and
 % level n uses a tail approximation via the G matrix.
 %
 % @par Syntax:
 % @code
 % D = qbd_depproc_etaqa(MAPa, MAPs, n)
 % @endcode
 %
 % @par Parameters:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>MAPa<td>Arrival process in MAP format {D0, D1}
 % <tr><td>MAPs<td>Service process in MAP format {D0, D1}
 % <tr><td>n<td>Truncation level (number of QBD levels to represent explicitly)
 % </table>
 %
 % @par Returns:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>D<td>Departure process in MAP format {D0, D1}
 % </table>
%}
function [D]=qbd_depproc_etaqa(MAPa,MAPs,n)

na = length(MAPa{1});
ns = length(MAPs{1});
lvlsz = ns*na;

F = kron(MAPa{2},eye(ns));
L = krons(MAPa{1},MAPs{1});
B = kron(eye(na),MAPs{2});
L0 = kron(MAPa{1},eye(ns));

[~,R,~] = QBD_CR(B,L,F);
G = inv(-L-R*B)*B;
Lhat = F+L;
Bbar = B+F*G;
Bhat = F*G;

vn1 = ones(1,n-1);
vn = zeros(1,n); vn(end)=1;

D0=kron(diag([1,vn1]),L)+kron(diag(vn1,1),F);
D0=[zeros(size(L0,1),size(D0,2)); D0];
D0 = [zeros(size(D0,1),size(B,2)),D0];
D0(1:size(L0,1),1:(size(L0,2)+size(F,2)))=[L0,F];
D0(((n-1)*lvlsz+1):n*lvlsz,((n-1)*lvlsz+1):n*lvlsz)=Lhat;

D1=kron(diag(vn,-1),Bbar)+kron(diag([0,vn]),Bhat);
D1(1:n*lvlsz,1:n*lvlsz)=kron(diag(vn1,-1),B);
D={D0,D1};
end
